doc/source/user-profiling.rst

```diff
@@ -0,0 +1,486 @@
+function wrapper to call ``time()`` right before and right after each loop 
+function to print out how long each loop takes overall:
+
+.. code-block:: python
+
+  # This is a generic wrapper for any driver function you want to time
+  def time_this(f):
+    def timed_wrapper(*args, **kw):
+      start_time = time.time()
+      result = f(*args, **kw)
+      end_time = time.time()
+
+      # Time taken = end_time - start_time
+      print('| func:%r args:[%r, %r] took: %2.4f seconds |' % \
+            (f.__name__, args, kw, end_time - start_time))
+      return result
+    return timed_wrapper
+
+To **always** print out how long the loop takes to run each time the loop 
+function ``ex1()`` is called, we can evoke our ``time_this`` wrapper with 
+a function decorator. This can similarly be done to functions ``ex2()``
+and ``ex3()``:
+
+.. code-block:: python
+
+  @time_this  # Added decorator
+  def ex1():
+    list1 = []
+    for i in range(5):
+      list1.append(ray.get(func.remote()))
+
+  def main():
+    ray.init()
+    ex1()
+    ex2()
+    ex3()
+
+  if __name__ == "__main__":
+    main()
+
+Alternatively, to print out the timer on **selective** calls to ``ex1()``,
+we can forgo the decorator and make explicit calls using ``time_this``
+as follows:
+
+.. code-block:: python
+
+  def ex1():  # Removed decorator
+    list1 = []
+    for i in range(5):
+      list1.append(ray.get(func.remote()))
+
+  def main():
+    ray.init()
+    ex1()             # This call outputs nothing
+    time_this(ex1)()  # This call outputs total execution time of ex1
+    time_this(ex2)()
+    time_this(ex3)()
+
+  if __name__ == "__main__":
+    main()
+
+Finally, running the three timed loops should yield output similar to this:
+
+.. code-block:: bash
+
+  | func:'ex1' args:[(), {}] took: 2.5083 seconds |
+  | func:'ex2' args:[(), {}] took: 1.0032 seconds |
+  | func:'ex3' args:[(), {}] took: 1.1045 seconds |
+
+Let's interpret these results. 
+
+Most pertinently, ``ex1()`` took substantially more time than ``ex2()``, 
+despite their only difference being that ``ex1()`` calls ``ray.get`` on the 
+remote function before adding it to the list, while ``ex2()`` waits to fetch 
+the entire list with ``ray.get`` at once.
+
+.. code-block:: python
+
+  @ray.remote
+  def func(): # A single call takes 0.5 seconds
+    time.sleep(0.5)
+
+  def ex1():  # Took Ray 2.5 seconds
+  list1 = []
+  for i in range(5):
+    list1.append(ray.get(func.remote()))
+
+  def ex2():  # Took Ray 1 second
+    list2 = []
+    for i in range(5):
+      list2.append(func.remote())
+    ray.get(list2)
+
+Notice how ``ex1()`` took 2.5 seconds, exactly five times 0.5 seconds, or 
+the time it would take to wait for our remote function five times in a row. 
+
+By calling ``ray.get`` after each call to the remote function, ``ex1()`` 
+removes all ability to parallelize work, by forcing the driver to wait for 
+each ``func()``'s result in succession. We are completely sabotaging any 
```

Comment:
Instead of sabotaging, let's say "we are not taking advantage of Ray parallelization here" :)